src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,877 @@
+        );
+    }
+    debug_assert!(
+        part_with_delta.first_key().is_some(),
+        "must have something in the range cache after applying delta"
+    );
+
+    // Let's think about the following algorithm with the following cases in mind.
+    // Insertions are relatively easy, so we only give the deletion examples.
+    // The two directions are symmetrical, we only consider the left direction in
+    // the following description.
+    //
+    // ## Background
+    //
+    // *Sm* means smallest sentinel node in the cache
+    // *SM* means largest sentinel node in the cache
+    //
+    // Before calling this function, all entries within range covered by delta should
+    // have been loaded into the cache.
+    //
+    // ## Cases
+    //
+    // Frame: ROWS BETWEEN ? AND 2 FOLLOWING
+    // Delta: Delete 5
+    // Cache + Delta:
+    //   [1, 2, 6] -> 1
+    //   [1, 6] -> 1
+    //   [6, 7, 8] -> 6, not precise but won't do too much harm
+    //   [1, 2] -> 1
+    //   [1] -> 1
+    //   [Sm, 6] -> Sm
+    //   [Sm, 1] -> Sm
+    //   [1, 2, SM] -> 1
+    //   [1, SM] -> 1
+    //   [Sm, 1, SM] -> Sm
+    //   [Sm, SM] -> Sm
+    //
+    // Frame: ROWS BETWEEN ? AND CURRENT ROW
+    // Delta: Delete 5
+    // Cache + Delta:
+    //   [1, 2, 6] -> 6, not precise but won't do too much harm
+    //   [1, 2] -> 2, not precise but won't do too much harm
+    //   [1, 2, SM] -> SM, not precise but won't do too much harm
+    //   [Sm, SM] -> SM, not precise but won't do too much harm
+    //
+    // Frame: ROWS BETWEEN ? AND 2 PRECEDING
+    // This will be treated as if it's `ROWS BETWEEN ? AND CURRENT ROW`.
+
+    let mut cursor = if LEFT {
+        part_with_delta.lower_bound(Bound::Included(delta_key))
+    } else {
+        part_with_delta.upper_bound(Bound::Included(delta_key))
+    };
+    let n_rows_to_move = if LEFT {
+        frame_bounds.n_following_rows().unwrap()
+    } else {
+        frame_bounds.n_preceding_rows().unwrap()
+    };
+
+    if n_rows_to_move == 0 {
+        return cursor
+            .key()
+            .or_else(|| {
+                if LEFT {
+                    part_with_delta.last_key()
+                } else {
+                    part_with_delta.first_key()
+                }
+            })
+            .unwrap();
+    }
+
+    for _ in 0..n_rows_to_move {
+        // Note that we have to move before check, to handle situation where the
+        // cursor is at ghost position at first.
+        if LEFT {
+            cursor.move_prev();
+        } else {
+            cursor.move_next();
+        }
+        if cursor.position().is_ghost() {
+            break;
+        }
+    }
+    cursor
+        .key()
+        .or_else(|| {
+            // Note the difference between this with the `n_rows_to_move == 0` case.
+            if LEFT {
+                part_with_delta.first_key()
+            } else {
+                part_with_delta.last_key()
+            }
+        })
+        .unwrap()
```

Comment:
Thinking about it further, perhaps it will simplify it greatly if we just inlined these inside `find_first_curr` and `find_last_curr`.

Because here the logic is mostly branched on `LEFT`, which makes it unreadable.

Reply:
Actually I put them together deliberately🤣, so that we can compare the two direction directly, making sure that they are symmetric. We should think the two direction simultaneously. Previously they are separated and I found it hard to maintain correctness. By putting them in one place it's very easy to find bug like:

```rust
    for _ in 0..n_rows_to_move {
        // Note that we have to move before check, to handle situation where the
        // cursor is at ghost position at first.
        if LEFT {
            cursor.move_prev();
        } else {
            cursor.move_prev();
        }
        if cursor.position().is_ghost() {
            break;
        }
    }
```